openedx/core/djangoapps/programs/utils.py

```diff
@@ -127,6 +128,45 @@ def engaged_programs(self):
 
         return programs
 
+    def _is_course_in_progress(self, now, enrolled_run_modes, course):
+        """Check if course qualifies as in progress as part of the program.
+
+        A course is considered to be in progress if a user is enrolled in a run
+        of the correct mode or a run of the correct mode is still available for enrollment.
+
+        Arguments:
+            now (datetime): datetime for now
+            enrolled_run_modes (dict): Dict showing what mode is enrolled for each run.
+            course (dict): Containing nested course runs.
+
+        Returns:
+            bool, indicating whether the course is in progress.
+        """
+        # Part 1: Check if any of the seats you are enrolled in qualify this course as in progress
+        enrolled_runs = [run for run in course['course_runs'] if run['key'] in self.course_run_ids]
+        # Check if the user is enrolled in the required mode for the run
+        runs_with_required_mode = [
+            run for run in enrolled_runs
+            if run.get('type') == enrolled_run_modes[run['key']]
```

Comment:
That's odd, the `CourseRunFactory` does define `type`. You shouldn't have to touch the factory to add the key.

I wouldn't let the tests influence the way you write the code. Every course run that's part of a program must have a type, otherwise we have no way of knowing when it's been completed. The discovery service has to include this key. If it doesn't, things should blow up so we know there's a problem.